lib/pdk/util/bundler.rb

```diff
@@ -22,19 +25,33 @@ def self.ensure_bundle!(gem_overrides = {})
           return
         end
 
-        # Generate initial Gemfile.lock
-        if bundle.locked?
-          # Update puppet-related gem dependencies by re-resolving them specifically.
-          # If this is a packaged install, only consider already available gems at this point.
-          bundle.update_lock!(gem_overrides, local: PDK::Util.package_install?)
-        else
-          bundle.lock!(gem_overrides)
+        unless bundle.locked?
+          # Generate initial default Gemfile.lock, either from package cache or
+          # by invoking `bundle lock`
+          bundle.lock!
+        end
+
+        # Check if all dependencies will be available once we update the lockfile.
+        begin
+          original_lockfile = bundle.gemfile_lock
+          temp_lockfile = "#{original_lockfile}.tmp"
+
+          FileUtils.mv(original_lockfile, temp_lockfile)
+
+          all_deps_available = bundle.installed?(gem_overrides)
+        ensure
+          FileUtils.mv(temp_lockfile, original_lockfile, force: true)
         end
 
-        # Check for any still-unresolved dependencies. For packaged installs, this should
-        # only evaluate to false if the user has added custom gems that we don't vendor, in
-        # which case `bundle install` will resolve new dependencies as needed.
-        unless bundle.installed?(gem_overrides)
+        # Update puppet-related gem dependencies by re-resolving them specifically.
+        # If there are additional dependencies that aren't available locally, allow
+        # `bundle lock` to reach out to rubygems.org
+        bundle.update_lock!(gem_overrides, local: all_deps_available)
+
+        # If there were missing dependencies when we checked above, let `bundle install`
+        # go out and get them. For packaged installs, this should only be true if the user
+        # has added custom gems that we don't vendor.
+        unless bundle.installed?
```

Comment:
Hmm, that's odd. If `all_deps_available` was true then `bundle lock --update` shouldn't reach out to rubygems for updates since we are passing the `--local` flag. This shouldn't hurt anything though but it is curious.